common/quotas/health_request_rate_limiter_impl.go

```diff
@@ -0,0 +1,116 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package quotas
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"go.temporal.io/server/common/backpressure"
+)
+
+type (
+	HealthRequestRateLimiterImpl struct {
+		rateLimiter   *RateLimiterImpl
+		healthSignals backpressure.HealthSignalAggregator
+
+		latencyThreshold    float64
+		errorRatioThreshold float64
+
+		refreshTimer    *time.Timer
+		refreshInterval time.Duration
+
+		// reductionMultiplier will be used to get the new rate limit by multiplying it with the current rate
```

Comment:
on what conditions does it do that?

can you write a description of the overall algorithm in a big comment somewhere?